File: Piko/Sources/SFMixerNode.m
//
//  SFMixerNode.m
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

#import "SFMixerNode.h"

#import "SFBuffer.h"
#import "SFBufferOperator.h"

@implementation SFMixerNode

- (instancetype)init
{
    SFInput *i1 = [[SFInput alloc] initWithName:@"1"];
    SFInput *i2 = [[SFInput alloc] initWithName:@"2"];
    SFInput *i3 = [[SFInput alloc] initWithName:@"3"];
    SFInput *i4 = [[SFInput alloc] initWithName:@"4"];
    SFOutput *output = [[SFOutput alloc] initWithName:@"output"];
    self = [super initWithInputs:@[i1, i2, i3, i4] outputs:@[output]];
    if (self != nil) {
        _input1 = i1;
        _input2 = i2;
        _input3 = i3;
        _input4 = i4;
        _output = output;
    }
    return self;
}

- (BOOL)rendersConstantFrame
{
    return NO;
}

- (BOOL)renderForPlayer:(SFPlaybackManager *)player
                  index:(UInt64)index
                  count:(NSInteger)count
                     to:(double *)buffer
{
    BOOL didRender = NO;
    SFBufferOperator *tempBufferOperator;

    for (SFInput *input in self.inputs) {
        SFNode *node = input.source.owner;
        if (node == nil) {
            continue;
        }

        if ([node rendersConstantFrame]) {
            double sample;
            if (![node renderForPlayer:player index: index count: 1 to: &sample]) {
                return NO;
            }
            if (didRender) {
                for (NSInteger i = 0; i < count; i++) {
                    buffer[i] += sample;
                }
            } else {
                for (NSInteger i = 0; i < count; i++) {
                    buffer[i] = sample;
                }
            }
        } else {
            if (didRender) {
                if (tempBufferOperator == nil) {
                    tempBufferOperator = [player getSFBufferOperator];
                }
                double *tempBufferData = tempBufferOperator.buffer.data;

                if (![node renderForPlayer:player index:index count:count to:tempBufferData]) {
                    return NO;
                }
                for (NSInteger i = 0; i < count; i++) {
                    buffer[i] += tempBufferData[i];
                }
            } else {
                if (![node renderForPlayer:player index:index count:count to:buffer]) {
                    return NO;
                }
            }
        }
        didRender = YES;
    }
    return didRender;
}

@end

File: Piko/Sources/SFMixerNode.swift
//
//  SFMixerNode.swift
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

import Foundation

@objc class SFMixerNode: SFNode {

    let input1: SFInput
    let input2: SFInput
    let input3: SFInput
    let input4: SFInput
    let output: SFOutput

    init() {
        input1 = SFInput(name: "1")
        input2 = SFInput(name: "2")
        input3 = SFInput(name: "3")
        input4 = SFInput(name: "4")
        output = SFOutput(name: "output")
        super.init(inputs: [input1, input2, input3, input4], outputs: [output])
    }

    override func renderForPlayer(player: SFPlaybackManager,
                                   index: UInt64,
                                   count: Int,
                               to buffer: UnsafeMutablePointer<Double>) -> Bool {

        var didRender: Bool = false
        var tempBufferOperator: SFBufferOperator?

        for input in inputs {
            if input.source == nil {
                continue
            }

            let node = input.source!.owner!

            if node.rendersConstantFrame() {
                var sample: Double = 0.0
                if node.renderForPlayer(player, index: index, count: 1, to: &sample) == false {
                    return false
                }
                if didRender {
                    for i in 0..<count {
                        buffer[i] += sample
                    }
                } else {
                    for i in 0..<count {
                        buffer[i] = sample
                    }
                }
            } else {
                if didRender {
                    if tempBufferOperator == nil {
                        tempBufferOperator = player.getSFBufferOperator()
                    }
                    let tempBufferData = tempBufferOperator!.buffer.data

                    if node.renderForPlayer(player, index: index, count: count, to: tempBufferData) == false {
                        return false
                    }
                    for i in 0..<count {
                        buffer[i] += tempBufferData[i]
                    }
                } else {
                    if node.renderForPlayer(player, index: index, count: count, to: buffer) == false {
                        return false
                    }
                }
            }
            didRender = true
        }
        return didRender
    }
}
